### ISSUE TITLE
Check for call Close for sql.Rows data type

### ISSUE THREAD

**Issue by UncleAndy** [go-critic/go-critic#430]
This is critical error, if not called Close for sql.Rows data.

But this can happen in different situations:

1. Simple situation:
`rows, err := db.Query(...)
for rows.Next { ... }
rows.Close`

2. Close in defer:
`rows, err := db.Query(...)
if err != nil {...}
defer rows.Close
for rows.Next { ... }`

3. Close in other method:
`rows, err := db.Query(...)
if err != nil {...}
rowsProcessor(rows)

func rowsProcessor(rows *sql.Rows) {
  ...
  rows.Close
}`

I wand do it self, but I not sure about gocritic posibility.

**Comment by Quasilyte** [comment]
Interesting.
The case with `Close` in other methods is the most challenging.
As a temporary workaround, we may give up analysis if tracked rows object escapes to other function.

For cases when rows do not escape their scope (I think this case is quite common),
we can track all `*sql.Rows` that are created.

It could be like this:
* Walk function body, for every call like `sql.Query` track returned rows object by adding it to tracked objects set (or list, whatever).
* If found any function call that takes any of tracked objects as an argument, remove that object from tracked list.
* For every `Close` method call remove receiver from the tracked objects set.

The only problem left is how to ensure that Close is called inside correct execution path.
Simplest solution is to require it to be called unconditionally, inside the same block that created these rows. Control flow analysis over AST would be harder than on SSA form, thing to keep in mind.

**Comment by cristaloleg** [comment]
@UncleAndy do you want to start working on this? 😉

**Comment by UncleAndy** [comment]
> @UncleAndy do you want to start working on this?

Yes, of couse. But I see gocritic today first time. It takes time.

**Comment by cristaloleg** [comment]
Sure, no rush.

**Comment by Quasilyte** [comment]
@UncleAndy, we've opened telegram chat just now, to make communication easier:
https://t.me/joinchat/DWka6kba5sa_EwTgmd3Vng (RU)
https://t.me/joinchat/DWka6g9VbCADtJI5J5w8nQ (EN)
Feel free to join and ask any questions.

**Comment by Antonboom** [comment]
Related to https://github.com/golangci/golangci-lint/pull/3348#issuecomment-1742046805
